diff -ur dev_gvisor_gvisor.orig/pkg/sentry/kernel/syslog.go dev_gvisor_gvisor/pkg/sentry/kernel/syslog.go
--- dev_gvisor_gvisor.orig/pkg/sentry/kernel/syslog.go	2023-03-22 18:31:40.976045180 +0100
+++ dev_gvisor_gvisor/pkg/sentry/kernel/syslog.go	2023-03-22 18:40:18.143912568 +0100
@@ -81,6 +81,11 @@
 		"Recruiting cron-ies...",
 		"Verifying that no non-zero bytes made their way into /dev/zero...",
 		"Accelerating teletypewriter to 9600 baud...",
+		"Supervising the supervisor...",
+		"Adding more KubernetesWorker nodes...",
+		"Picking a new name for Metropolis...",
+		"Re-reading the TPM spec...",
+		"Re-implementing mkfs.xfs in Go...",
 	}
 
 	selectMessage := func() string {
@@ -96,7 +101,7 @@
 
 	const format = "<6>[%11.6f] %s\n"
 
-	s.msg = append(s.msg, []byte(fmt.Sprintf(format, 0.0, "Starting gVisor..."))...)
+	s.msg = append(s.msg, []byte(fmt.Sprintf(format, 0.0, "Starting gVisor on Metropolis..."))...)
 
 	time := 0.1
 	for i := 0; i < 10; i++ {
