From ba6c6c23295a765158a2e6d94879173419b72926 Mon Sep 17 00:00:00 2001
From: Lorenz Brun <lorenz@brun.one>
Date: Sat, 30 Nov 2024 01:03:23 +0100
Subject: [PATCH 1/2] Increase netlink socket buffers

---
 conn.go | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/conn.go b/conn.go
index fef9c2a..a315db3 100644
--- a/conn.go
+++ b/conn.go
@@ -73,6 +73,8 @@ func New(opts ...ConnOption) (*Conn, error) {
 	if err != nil {
 		return nil, err
 	}
+	nlconn.SetReadBuffer(1 << 22)
+	nlconn.SetWriteBuffer(1 << 22)
 	cc.nlconn = nlconn
 	return cc, nil
 }
-- 
2.47.0

