From e41145bf0e32b404477ec89f76fb99fe1c33874a Mon Sep 17 00:00:00 2001
From: Lorenz Brun <lorenz@monogon.tech>
Date: Wed, 4 May 2022 18:47:00 +0200
Subject: [PATCH] Bump Kubernetes to 1.24

diff --git a/plugin/kubernetes/object/object.go b/plugin/kubernetes/object/object.go
index 7111833e..34217797 100644
--- a/plugin/kubernetes/object/object.go
+++ b/plugin/kubernetes/object/object.go
@@ -100,11 +100,11 @@ func (e *Empty) GetOwnerReferences() []v1.OwnerReference { return nil }
 // SetOwnerReferences implements the metav1.Object interface.
 func (e *Empty) SetOwnerReferences([]v1.OwnerReference) {}
 
-// GetClusterName implements the metav1.Object interface.
-func (e *Empty) GetClusterName() string { return "" }
+// GetZZZ_DeprecatedClusterName implements the metav1.Object interface.
+func (e *Empty) GetZZZ_DeprecatedClusterName() string { return "" }
 
-// SetClusterName implements the metav1.Object interface.
-func (e *Empty) SetClusterName(clusterName string) {}
+// SetZZZ_DeprecatedClusterName implements the metav1.Object interface.
+func (e *Empty) SetZZZ_DeprecatedClusterName(clusterName string) {}
 
 // GetManagedFields implements the metav1.Object interface.
 func (e *Empty) GetManagedFields() []v1.ManagedFieldsEntry { return nil }
-- 
2.25.1

