From 49b5f957d01cd50e5303cfc0e6c7f350d7e1c24e Mon Sep 17 00:00:00 2001
From: Lorenz Brun <lorenz@monogon.tech>
Date: Tue, 30 Nov 2021 03:19:37 +0100
Subject: [PATCH] Alias go_default_library in BUILD to make Gazelle work

---
 BUILD.bazel | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/BUILD.bazel b/BUILD.bazel
index 9a1edd4..e5b8f96 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -14,6 +14,12 @@ gazelle(
 # By default resolve the top-level package to the gogo variant as it's required for the plugin compilation.
 # gazelle:resolve go github.com/mwitkow/go-proto-validators //:validators_gogo
 
+alias(
+    name = "go_default_library",
+    actual = "validators_golang",
+    visibility = ["//visibility:public"],
+)
+
 go_proto_compiler(
     name = "go_proto_validators",
     options = ["gogoimport=false"],
-- 
2.25.1

