--- a/pkg/sentry/platform/kvm/BUILD.bazel
+++ b/pkg/sentry/platform/kvm/BUILD.bazel
@@ -6,6 +6,7 @@
         "address_space.go",
         "address_space_amd64.go",
         "address_space_arm64.go",
+        "address_space_debug.go",
         "atomicptr_machine_unsafe.go",
         "bluepill.go",
         "bluepill_allocator.go",
@@ -15,6 +16,7 @@
         "bluepill_arm64.go",
         "bluepill_arm64.s",
         "bluepill_arm64_unsafe.go",
+        "bluepill_debug_unsafe.go",
         "bluepill_fault.go",
         "bluepill_unsafe.go",
         "context.go",
@@ -40,6 +42,7 @@
         "machine_amd64_unsafe.go",
         "machine_arm64.go",
         "machine_arm64_unsafe.go",
+        "machine_debug.go",
         "machine_unsafe.go",
         "physical_map.go",
         "physical_map_amd64.go",