Fix for https://github.com/google/gvisor/issues/10483

--- a/pkg/xdp/xdp.go
+++ b/pkg/xdp/xdp.go
@@ -151,9 +151,9 @@
 	}
 
 	reg := unix.XDPUmemReg{
-		Addr: uint64(sliceBackingPointer(umemMemory)),
-		Len:  uint64(len(umemMemory)),
-		Size: opts.FrameSize,
+		Addr:       uint64(sliceBackingPointer(umemMemory)),
+		Len:        uint64(len(umemMemory)),
+		Chunk_size: opts.FrameSize,
 		// Not useful in the RX path.
 		Headroom: 0,
 		// TODO(b/240191988): Investigate use of SHARED flag.
