diff --git a/pkg/registry/core/service/portallocator/controller/BUILD.bazel b/pkg/registry/core/service/portallocator/controller/BUILD.bazel
index fa870760490..0f67fe7eb9b 100644
--- a/pkg/registry/core/service/portallocator/controller/BUILD.bazel
+++ b/pkg/registry/core/service/portallocator/controller/BUILD.bazel
@@ -26,6 +26,7 @@
         "@io_k8s_client_go//util/retry",
         "@io_k8s_component_base//metrics",
         "@io_k8s_component_base//metrics/legacyregistry",
+        "@@//metropolis/node",
     ],
 )

