From 5cf776eb8f872f78cf7325c39fbfe1777f702407 Mon Sep 17 00:00:00 2001
From: Mike Danese <mikedanese@google.com>
Date: Fri, 26 Feb 2021 13:39:16 -0800
Subject: [PATCH] fix go_genrule for 1.16

---
 defs/go.bzl | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/defs/go.bzl b/defs/go.bzl
index f4617e1..8873253 100644
--- a/defs/go.bzl
+++ b/defs/go.bzl
@@ -90,6 +90,9 @@ def _go_genrule_impl(ctx):
         "PATH": ctx.configuration.host_path_separator.join(["/usr/local/bin", "/bin", "/usr/bin"]),
         "GOPATH": paths.dirname(gopath_placeholder.path),
         "GOROOT": paths.dirname(go.sdk.root_file.path),
+        # hack to tie us over until we fix this to use modules or stop using
+        # it.
+        "GO111MODULE": "off",
     })
 
     ctx.actions.run(
